signature/src/mechanism/cosign/mod.rs

```diff
@@ -0,0 +1,346 @@
+pub const COSIGN_KEY_KBS: &str = "Cosign Key";
+
+#[derive(Deserialize, Debug, Eq, PartialEq, Serialize, Default)]
+pub struct CosignParameters {
+    // KeyPath is a pathname to a local file containing the trusted key(s).
+    // Exactly one of KeyPath and KeyData can be specified.
+    //
+    // This field is optional.
+    #[serde(rename = "keyPath")]
+    pub key_path: Option<String>,
+    // KeyData contains the trusted key(s), base64-encoded.
+    // Exactly one of KeyPath and KeyData can be specified.
+    //
+    // This field is optional.
+    #[serde(rename = "keyData")]
+    pub key_data: Option<String>,
+
+    // SignedIdentity specifies what image identity the signature must be claiming about the image.
+    // Defaults to "match-exact" if not specified.
+    //
+    // This field is optional.
+    #[serde(default, rename = "signedIdentity")]
+    pub signed_identity: Option<PolicyReqMatchType>,
+}
+
+#[async_trait]
+impl SignScheme for CosignParameters {
+    /// This initialization will:
+    /// * Create [`COSIGN_KEY_DIR`] if not exist.
+    async fn init(&self) -> Result<()> {
+        if !Path::new(COSIGN_KEY_DIR).exists() {
+            fs::create_dir_all(COSIGN_KEY_DIR).await.map_err(|e| {
+                anyhow!("Create Simple Signing sigstore-config dir failed: {:?}", e)
+            })?;
+        }
+
+        Ok(())
+    }
+
+    fn resource_manifest(&self) -> HashMap<&str, &str> {
+        let mut manifest = HashMap::new();
+        if let Some(key_path) = &self.key_path {
+            if !Path::new(key_path).exists() {
+                manifest.insert(COSIGN_KEY_KBS, &key_path[..]);
+            }
+        }
+        manifest
+    }
+
+    /// Judge whether an image is allowed by this SignScheme.
+    async fn allows_image(&self, image: &mut Image) -> Result<()> {
+        // Check before we access the network
+        self.check_reference_rule_types()?;
+
+        // Verification, will access the network
+        let payloads = self.verify_signature_and_get_payload(image).await?;
+
+        // check the reference rules (signed identity)
+        for payload in payloads {
+            if let Some(rule) = &self.signed_identity {
+                payload.validate_signed_docker_reference(&image.reference, rule)?;
+            }
+
+            payload.validate_signed_docker_manifest_digest(&image.manifest_digest.to_string())?;
+        }
+
+        Ok(())
+    }
+}
+
+impl CosignParameters {
+    /// Check whether this Policy Request Match Type (i.e., signed identity
+    /// check type) for the reference
+    /// is MatchRepository or ExactRepository.
+    /// Because cosign-created signatures only contain a repository,
+    /// so only matchRepository and exactRepository can be used to accept them.
+    /// Other types are all to be denied.
+    /// If it is neither of them, return `Error`. Otherwise, return `Ok()`
+    fn check_reference_rule_types(&self) -> Result<()> {
```

Comment:
Could you add a unit test for all the rule scenarios?

Reply:
Yes, will add it the next day

Reply:
Added